src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs

```diff
@@ -323,20 +323,36 @@ private void Init(string path, FileMode mode, FileAccess access, FileShare share
                 }
             }
 
-            // If preallocationSize has been provided for a creatable and writeable file
-            if (FileStreamHelpers.ShouldPreallocate(preallocationSize, access, mode, this))
+            if (preallocationSize > 0)
             {
-                int fallocateResult = Interop.Sys.PosixFAllocate(this, 0, preallocationSize);
-                if (fallocateResult != 0)
+                if (Interop.Sys.FAllocate(this, 0, preallocationSize) < 0)
                 {
-                    Dispose();
-                    Interop.Sys.Unlink(path!); // remove the file to mimic Windows behaviour (atomic operation)
-
-                    Debug.Assert(fallocateResult == -1 || fallocateResult == -2);
-                    throw new IOException(SR.Format(
-                        fallocateResult == -1 ? SR.IO_DiskFull_Path_AllocationSize : SR.IO_FileTooLarge_Path_AllocationSize,
-                        path,
-                        preallocationSize));
+                    Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
+
+                    if (errorInfo.Error != Interop.Error.EOPNOTSUPP && // Not supported by filesystem.
+                        errorInfo.Error != Interop.Error.ENOSYS &&     // Not supported by kernel.
+                        errorInfo.Error != Interop.Error.ESPIPE &&     // Not a file.
+                        errorInfo.Error != Interop.Error.ENODEV)       // Not a file.
+                    {
+                        Dispose();
+                        Interop.Sys.Unlink(path!); // remove the file to mimic Windows behaviour (atomic operation)
+
+                        switch (errorInfo.Error)
+                        {
+                            case Interop.Error.EFBIG:
+                                throw new IOException(SR.Format(
+                                    SR.IO_FileTooLarge_Path_AllocationSize,
+                                    path,
+                                    preallocationSize));
+                            case Interop.Error.ENOSPC:
+                                throw new IOException(SR.Format(
+                                    SR.IO_DiskFull_Path_AllocationSize,
+                                    path,
+                                    preallocationSize));
+                            default:
+                                throw Interop.GetExceptionForIoErrno(errorInfo, path, isDirectory: false);
```

Comment:
If preallocationSize is really meant as a hint, do we need to throw for such cases?

Reply:
Yes, I think it makes sense to not throw here. I'll change this.

Reply:
To be clear: I'm taking about the `default` case. I think it makes sense to throw in the other cases to avoid a user getting an exception later when he runs out of space while writing the file.

Reply:
> To be clear: I'm taking about the default case

I was, too :)

Reply:
Same as my previous comment, if we are not going to throw, we should not Dispose nor call Unlink, right?